### ISSUE TITLE
Benchmarks generating large number of files

### ISSUE THREAD

**Issue by gaow** [stephenslab/dsc2#29]
This is a challenge: @dcgerard has a benchmark where multiple replicates are generated (via `seed`) along with various combinations of parameters in computational routines. The DSC sequence involves multiple steps, too, i.e. multiple scenario + method + score. DSC keeps intermediate results at each computational step. So the total number of files from all steps is over 200,000. 

Basically if you add 2 more parameters at the first step, it will cause a "chain reaction", an exponential increase in the number of files.

It takes DSC2 90 minutes just to prepare the file signatures for the 200,000 potential outcomes before running the benchmark. I can only imaging bigger problems when DSC2 tries to load every of the 200,000 data-sets to create the final query database. Some smart engineering is needed here.

**Comment by gaow** [comment]
I should mention that currently it's a 106MB file just to save the MD5 summary of the steps along with the particular parameter combinations, for David's benchmark!

**Comment by gaow** [comment]
More thoughts here:

## impossibles
* DSC, by its design philosophy, keeps track of all intermediate files. It is impossible to compromise on this.
* It is also impossible to replace RDS format with other formats for intermediate data storage. RDS is the only decent format that can fully support arbitrary R object. Unlike HDF5, RDS files does not allow parallel I/O, incremental storage and partially access. So basically it is impossible to get away with not having to write millions of small files in a DSC benchmark. 

## possibles
* In preparing file signature it is possible to utilize some database that supports parallel I/O and some data compression that results in a file signature database 1) smaller and 2) easiler to query than the current implementation (text database, 206MB plain text for David's benchmark). SQLite is no good here. Looks like [redis](http://redis.io/) might be something to try out. (I have no experience with it).
* In putting together the result database, it is also possible to parse all RDS in parallel and store to some temporary format (HDF5, or redis again for its parallel support?) and convert to RDS file after the database is built. Here we can use other intermediate formats because the final database will be plain matrices / dataframe and I've already written a Python interface for handling them.

It's going to be some efforts to make these happen. But I should have seen this coming ... any DSC benchmark can easily generate millions of files.

**Comment by road2stat** [comment]
Cross posting from Slack:

- I agree with you about the two impossibles.

To me storing 20,0000 files in a good file system (ext4) is ok (maybe need some hiercharchical folder structures), but it does seem messy when things get to that scale.

My suggestion would be using a local file-based key-value database to store the RDS files and the keys to find them (this may also avoid manual hashing work since the unique keys here will be generated by the database?). Anyway, `gdbm` and `leveldb` can be good candidates.

Redis maybe has the best read/write performance across all databases, but it is primarily a memory-based db, which means the data all exist in the memory, and this can be soon unpractical if the intermediate files are huge.

**Comment by gaow** [comment]
Exactly -- Redis is a local key-value storage data base. But I have no intention to serialize the RDS and write them to database directly. The Redis is just a look up table for file signatures and it's going to be at most a few GB (to index a couple billions intermediate files). 

Are you suggesting I can use `gdbm` or `leveldb` to store the RDS files directly?

**Comment by stephens999** [comment]
I'd say there are two issues/approaches
i) you probably want to make filing issues as efficient as you can
ii) but in extreme cases, there is a big (and untenable?) 
overhead in saving all the intermediate files. The obvious
solution is not to save all intermediate files in some settings. You say that "by its design philosophy"
it is "impossible to compromise on this", but I'm not so sure. If the user requires, is it hard to think about
implementing the ability to run two steps together without saving intermediate results? Conceptually it seems to just be defining a "step" in the DSC as the sequential application of 2 steps.

**Comment by gaow** [comment]
>  If the user requires, is it hard to think about implementing the ability to run two steps together without saving intermediate results? 

Currently this is supported, via the `exec[1] + exec[2]` logic specification. For instance, instead of 

```yaml
simulate:
  exec: sim.R
  return: X

calculate:
  exec: calc.R
  params:
    X: $X
  return: Y
```

it is possible to write:

```
all_together:
   exec: sim.R, calc.R
   logic: exec[1] + exec[2]
   return: Y
```

However under the hood, I still save outcome from `exec[1]` (`sim.R`) for use with the next routine. The justification is that perhaps `exec[1]` is R script but `exec[2]` is Python or C executable. 

That said, it is possible to check the file extension and merge `exec[1]` and `exec[2]` to avoid saving intermediate files for this syntax; or introduce an "on the fly" option. I can try that. It will reduce number of files by at most 50% though.

Another thought is that we split DSC into multiple instances by its first step, because the first step does not have any dependencies the split will be most straightforward. For example when there are 1000 seeds * 10 parameter combinations we split the task into 100 mini, independent DSC. It will not reduce number of files saved, but it provides a most straightforward parallelazation to prepare file signatures; also it opens up the possibility to run DSC on cluster without having to implement a DAG (I'm not saying I'll not do DAG but this just seems easier and may get us a cluster version of DSC2 quicker). 

I can get both proposals to work.

**Comment by road2stat** [comment]
> Exactly -- Redis is a local key-value storage data base. But I have no intention to serialize the RDS and write them to database directly. The Redis is just a look up table for file signatures and it's going to be at most a few GB (to index a couple billions intermediate files).

I see, the db is used to store the keys only. The main reason I recommended `gdbm` is that Redis is client-server based, which means a system-level dependency that's not so easy to install for every platform, but indeed we can try them all out.

> Are you suggesting I can use gdbm or leveldb to store the RDS files directly? I'm not familiar with either of them. But a quick google search i did not see obvious support for parallel I/O (sees leveldb supports it in dev branch)

yes I guess you could store binary files in these dbs directly, although that's not necessary. As for disk i/o parallelization, that's a good point but might not influence a lot, just switching to some k-v dbs would be good enough for me.

**Comment by ramanshah** [comment]
The many-small-files problem is indeed a problem fairly globally - I've seen it cause problems in HDFS/Spark contexts (where it stresses the namenode) as well as in single-node ones.

Is `dsc2` a single-node or distributed system these days?

**Comment by gaow** [comment]
Hi @ramanshah great to hear from you! So currently since I've not implemented DAG for an upstream library that powers DSC2; for now it will run single-node even if jobs are submitted to cluster. It is on my to-do list. I've not yet considered HDFS/Spark, though. Perhaps it will not be hard to port it to that platform if there is a need, but we'll get DAG work first. 

Here is a summary of my plan: https://github.com/stephenslab/dsc2/issues/23

Thanks for your input!

**Comment by gaow** [comment]
Looked into redis. Performance-wise it is decent (without going multi-process) and data serialization is better compared to numpy. But a draw back is that it does require a redis server instance running, which might be a pain for users who are not software / computer savvy. 

Keep looking into other options ...

**Comment by ramanshah** [comment]
One thing that's an advantage of redis is that it's natively distributed. Glad to hear you're looking into it :)

**Comment by gaow** [comment]
So I'm taking another stab at this as I'm currently in need of the capability myself. Another issue that popped up is that in the preparation stage the memory consumption can be large in order to create filename mapping (file signature to human-readable filename). In a test it exceeded 8GB which maybe the limit for even a cluster thread. I need to even cache this step. I'm looking into [msgpack](http://msgpack.org/index.html) to efficiently caching the hash. Hopefully I can close this issue this week!

**Comment by gaow** [comment]
Currently using dictionary with msgpack caching, I'm able to obtain all file signatures for [David's example](https://github.com/dcgerard/sim_code/tree/master/dsc_sims/code) within 4 minutes. This is somewhat acceptable. Still trying to optimize some more.
